File: source/go/internal/jyotish/matching.go
package jyotish

import (
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/utils"
)

type NatalMoon struct {
	Rashi         int
	Nakshatra     int
	LagnaRashi    *int
	NavamsaRashi  *int
	NakshatraPada *int // validated, never read
}

type KootName string

const (
	KootVarna       KootName = "Varna"
	KootVashya      KootName = "Vashya"
	KootTara        KootName = "Tara"
	KootYoni        KootName = "Yoni"
	KootGrahaMaitri KootName = "Graha Maitri"
	KootGana        KootName = "Gana"
	KootBhakoot     KootName = "Bhakoot"
	KootNadi        KootName = "Nadi"
)

var AllKootNames = [8]KootName{
	KootVarna, KootVashya, KootTara, KootYoni,
	KootGrahaMaitri, KootGana, KootBhakoot, KootNadi,
}

var KootMaxScores = [8]float64{1, 2, 3, 4, 5, 6, 7, 8}

type AshtakootOptions struct {
	GanaCancellation bool // the reference almanac's 36-guna table applies no Gana cancellation
}

type KootScore struct {
	Name        KootName `json:"name"`
	Score       float64  `json:"score"` // float64 for Tara alone: 1.5 per auspicious direction
	MaxScore    float64  `json:"maxScore"`
	Description string   `json:"description"`
}

type AshtakootResult struct {
	TotalScore    float64     `json:"totalScore"`
	Koots         []KootScore `json:"koots"`
	Cancellations []string    `json:"cancellations"` // non-nil even when empty
}

func ComputeAshtakoot(boy, girl NatalMoon, options AshtakootOptions) (AshtakootResult, error) {
	if err := validateNatalMoon(boy, "boy"); err != nil {
		return AshtakootResult{}, err
	}
	if err := validateNatalMoon(girl, "girl"); err != nil {
		return AshtakootResult{}, err
	}

	cancellations := make([]string, 0, 3)

	koots := []KootScore{
		scoreVarna(boy, girl),
		scoreVashya(boy, girl),
		scoreTara(boy, girl),
		scoreYoni(boy, girl),
		scoreGrahaMaitri(boy, girl),
		scoreGana(boy, girl, &cancellations, options.GanaCancellation),
		scoreBhakoot(boy, girl, &cancellations),
		scoreNadi(boy, girl, &cancellations),
	}

	total := 0.0
	for _, k := range koots {
		total += k.Score
	}
	return AshtakootResult{TotalScore: total, Koots: koots, Cancellations: cancellations}, nil
}

func validateNatalMoon(m NatalMoon, label string) error {
	if m.Rashi < 0 || m.Rashi >= 12 {
		return types.Codef(types.ErrInvalidInput,
			"%s.rashi must be integer in [0, 11], got %d", label, m.Rashi)
	}
	if err := utils.AssertNakshatraIndex(m.Nakshatra, label+".nakshatra"); err != nil {
		return err
	}
	if m.LagnaRashi != nil && (*m.LagnaRashi < 0 || *m.LagnaRashi >= 12) {
		return types.Codef(types.ErrInvalidInput,
			"%s.lagnaRashi must be integer in [0, 11], got %d", label, *m.LagnaRashi)
	}
	if m.NavamsaRashi != nil && (*m.NavamsaRashi < 0 || *m.NavamsaRashi >= 12) {
		return types.Codef(types.ErrInvalidInput,
			"%s.navamsaRashi must be integer in [0, 11], got %d", label, *m.NavamsaRashi)
	}
	if m.NakshatraPada != nil && (*m.NakshatraPada < 1 || *m.NakshatraPada > 4) {
		return types.Codef(types.ErrInvalidInput,
			"%s.nakshatraPada must be integer in [1, 4], got %d", label, *m.NakshatraPada)
	}
	return nil
}

func scoreVarna(boy, girl NatalMoon) KootScore {
	boyVarna := RashiVarna[boy.Rashi]
	girlVarna := RashiVarna[girl.Rashi]
	score := 0.0
	if VarnaRank[boyVarna] >= VarnaRank[girlVarna] {
		score = 1
	}
	return KootScore{
		Name: KootVarna, Score: score, MaxScore: 1,
		Description: "Boy's varna (" + string(boyVarna) + ") vs girl's (" + string(girlVarna) + ")",
	}
}

func scoreVashya(boy, girl NatalMoon) KootScore {
	bv := RashiVashya[boy.Rashi]
	gv := RashiVashya[girl.Rashi]
	return KootScore{
		Name: KootVashya, Score: VashyaScore[VashyaIndex(bv)][VashyaIndex(gv)], MaxScore: 2,
		Description: string(bv) + " ↔ " + string(gv),
	}
}

func scoreTara(boy, girl NatalMoon) KootScore {
	dBoyToGirl := ((girl.Nakshatra-boy.Nakshatra+27)%27 + 1)
	dGirlToBoy := ((boy.Nakshatra-girl.Nakshatra+27)%27 + 1)
	remBoyToGirl := dBoyToGirl % 9
	remGirlToBoy := dGirlToBoy % 9
	auspiciousBoyToGirl := !inauspiciousTara(remBoyToGirl)
	auspiciousGirlToBoy := !inauspiciousTara(remGirlToBoy)
	score := 0.0
	if auspiciousBoyToGirl {
		score += 1.5
	}
	if auspiciousGirlToBoy {
		score += 1.5
	}
	return KootScore{
		Name: KootTara, Score: score, MaxScore: 3,
		Description: "Boy→Girl rem " + jsnum.FormatInt(int64(remBoyToGirl)) +
			" (" + auspiciousLabel(auspiciousBoyToGirl) + "); Girl→Boy rem " +
			jsnum.FormatInt(int64(remGirlToBoy)) + " (" + auspiciousLabel(auspiciousGirlToBoy) + ")",
	}
}

func inauspiciousTara(rem int) bool {
	for _, r := range InauspiciousTaraRemainders {
		if r == rem {
			return true
		}
	}
	return false
}

func auspiciousLabel(ok bool) string {
	if ok {
		return "auspicious"
	}
	return "inauspicious"
}

func scoreYoni(boy, girl NatalMoon) KootScore {
	by := NakshatraYoni[boy.Nakshatra]
	gy := NakshatraYoni[girl.Nakshatra]
	return KootScore{
		Name: KootYoni, Score: float64(YoniScore[YoniIndex(by)][YoniIndex(gy)]), MaxScore: 4,
		Description: string(by) + " ↔ " + string(gy),
	}
}

func scoreGrahaMaitri(boy, girl NatalMoon) KootScore {
	boyLord := RashiLord[boy.Rashi]
	girlLord := RashiLord[girl.Rashi]
	var score float64
	var description string
	if boyLord == girlLord {
		score = 5
		// the published string reads "graha 2", not the lord's name
		description = "Same rashi-lord (graha " + jsnum.FormatInt(int64(boyLord)) + "), full marks"
	} else {
		boyView := NaisargikaMaitri[boyLord][girlLord]
		girlView := NaisargikaMaitri[girlLord][boyLord]
		score = GrahaMaitriScore[MaitriIdx(boyView)][MaitriIdx(girlView)]
		description = "Boy lord views girl lord as " + maitriLabel(boyView) +
			"; girl lord views boy lord as " + maitriLabel(girlView)
	}
	return KootScore{Name: KootGrahaMaitri, Score: score, MaxScore: 5, Description: description}
}

func maitriLabel(v int) string {
	switch v {
	case 1:
		return "friend"
	case -1:
		return "enemy"
	}
	return "neutral"
}

func scoreGana(boy, girl NatalMoon, cancellations *[]string, applyCancellation bool) KootScore {
	bg := NakshatraGana[boy.Nakshatra]
	gg := NakshatraGana[girl.Nakshatra]
	score := float64(GanaScore[GanaIdx(bg)][GanaIdx(gg)])
	description := string(bg) + " ↔ " + string(gg)

	// score ≤ 1 is the doshic pair: Manushya-Rakshasa (0) or Deva-Rakshasa (1)
	if applyCancellation && score <= 1 {
		boyLord := RashiLord[boy.Rashi]
		girlLord := RashiLord[girl.Rashi]
		sameLord := boyLord == girlLord
		mutualFriend := NaisargikaMaitri[boyLord][girlLord] == 1 &&
			NaisargikaMaitri[girlLord][boyLord] == 1
		if sameLord || mutualFriend {
			score = 6
			reason := lordReason(sameLord)
			description += ", cancelled by " + reason
			*cancellations = append(*cancellations, "Gana: "+reason)
		}
	}

	return KootScore{Name: KootGana, Score: score, MaxScore: 6, Description: description}
}

func lordReason(sameLord bool) string {
	if sameLord {
		return "same rashi-lord"
	}
	return "mutual friendship of rashi-lords"
}

func scoreBhakoot(boy, girl NatalMoon, cancellations *[]string) KootScore {
	dBoyToGirl := ((girl.Rashi-boy.Rashi+12)%12 + 1)
	dGirlToBoy := ((boy.Rashi-girl.Rashi+12)%12 + 1)

	isDoshic := false
	for _, p := range BhakootDoshicDistances {
		if p[0] == dBoyToGirl && p[1] == dGirlToBoy {
			isDoshic = true
			break
		}
	}

	score := 7.0
	description := "Distance (" + jsnum.FormatInt(int64(dBoyToGirl)) + ", " +
		jsnum.FormatInt(int64(dGirlToBoy)) + ")"
	if isDoshic {
		score = 0
		description = "Doshic distance (" + jsnum.FormatInt(int64(dBoyToGirl)) + ", " +
			jsnum.FormatInt(int64(dGirlToBoy)) + ")"
	}

	if isDoshic {
		boyLord := RashiLord[boy.Rashi]
		girlLord := RashiLord[girl.Rashi]
		sameLord := boyLord == girlLord
		mutualFriend := NaisargikaMaitri[boyLord][girlLord] == 1 &&
			NaisargikaMaitri[girlLord][boyLord] == 1
		if sameLord || mutualFriend {
			score = 7
			reason := lordReason(sameLord)
			description += ", cancelled by " + reason
			*cancellations = append(*cancellations, "Bhakoot: "+reason)
		} else {
			if boy.LagnaRashi != nil && girl.LagnaRashi != nil {
				boyLagnaLord := RashiLord[*boy.LagnaRashi]
				girlLagnaLord := RashiLord[*girl.LagnaRashi]
				if boyLagnaLord == girlLagnaLord {
					score = 7
					description += ", cancelled by same lagna-lord"
					*cancellations = append(*cancellations, "Bhakoot: same lagna-lord")
				} else {
					boySeventhLord := RashiLord[(*boy.LagnaRashi+6)%12]
					girlSeventhLord := RashiLord[(*girl.LagnaRashi+6)%12]
					if boySeventhLord == girlSeventhLord {
						score = 7
						description += ", cancelled by same 7th-house lord"
						*cancellations = append(*cancellations, "Bhakoot: same 7th-house lord")
					}
				}
			}
			if score == 0 && boy.NavamsaRashi != nil && girl.NavamsaRashi != nil {
				if RashiLord[*boy.NavamsaRashi] == RashiLord[*girl.NavamsaRashi] {
					score = 7
					description += ", cancelled by same Navamsa lord"
					*cancellations = append(*cancellations, "Bhakoot: same Navamsa lord")
				}
			}
		}
	}
	return KootScore{Name: KootBhakoot, Score: score, MaxScore: 7, Description: description}
}

func scoreNadi(boy, girl NatalMoon, cancellations *[]string) KootScore {
	bn := NakshatraNadi[boy.Nakshatra]
	gn := NakshatraNadi[girl.Nakshatra]
	sameNadi := bn == gn
	score := 8.0
	if sameNadi {
		score = 0
	}
	description := string(bn) + " ↔ " + string(gn)

	if sameNadi {
		sameNakshatra := boy.Nakshatra == girl.Nakshatra
		sameRashi := boy.Rashi == girl.Rashi
		if sameNakshatra || sameRashi {
			score = 8
			reason := "same rashi"
			if sameNakshatra {
				reason = "same nakshatra"
			}
			description += ", cancelled by " + reason
			*cancellations = append(*cancellations, "Nadi: "+reason)
		}
	}
	return KootScore{Name: KootNadi, Score: score, MaxScore: 8, Description: description}
}

File: source/ts/src/jyotish/matching.ts
import {
  RASHI_VARNA, VARNA_RANK,
  RASHI_VASHYA, VASHYA_SCORE, vashyaIndex,
  NAKSHATRA_YONI, YONI_SCORE, yoniIndex,
  RASHI_LORD, NAISARGIKA_MAITRI, GRAHA_MAITRI_SCORE, maitriIdx,
  NAKSHATRA_GANA, GANA_SCORE, ganaIdx,
  NAKSHATRA_NADI,
  INAUSPICIOUS_TARA_REMAINDERS,
  BHAKOOT_DOSHIC_DISTANCES,
} from './matchingTables';
import { assertNakshatraIndex } from '../utils/validation';

export interface NatalMoon {
  /** Moon's rashi at birth (0 = Mesha … 11 = Meena). */
  rashi: number;
  /** Moon's nakshatra at birth (0 = Ashwini … 26 = Revati). */
  nakshatra: number;
  /** Lagna rashi, 0..11. Enables the same-lagna-lord and same-7th-lord Bhakoot cancellations when BOTH natives carry it. */
  lagnaRashi?: number;
  /** Moon's Navamsa (D9) rashi, 0..11. Enables the same-Navamsa-lord Bhakoot cancellation when BOTH natives carry it. */
  navamsaRashi?: number;
  /** Nakshatra pada, 1..4. Validated but not used in scoring. */
  nakshatraPada?: number;
}

export type KootName = 'Varna' | 'Vashya' | 'Tara' | 'Yoni' | 'Graha Maitri' | 'Gana' | 'Bhakoot' | 'Nadi';

/** Behaviour switches for {@link computeAshtakoot}; every flag defaults off. */
export interface AshtakootOptions {
  /** Restore a doshic Gana score (≤ 1) to 6 when the rashi-lords are the same graha or mutual friends; the reference almanac applies no Gana cancellation. */
  ganaCancellation?: boolean;
}

export interface KootScore {
  name: KootName;
  score: number;
  maxScore: number;
  description: string;
}

export interface AshtakootResult {
  /** Sum of the 8 koot scores, 0..36. */
  totalScore: number;
  /** Per-koot breakdown in canonical Brihat-Samhita order. */
  koots: KootScore[];
  /** Cancellations that were *applied* (raised a 0 to full marks). */
  cancellations: string[];
}

/**
 * Ashtakoot Guna Milan: the 36-point compatibility test between groom (vara) and
 * bride (vadhu), scored from each native's natal Moon rashi and nakshatra.
 */
export function computeAshtakoot(
  boy: NatalMoon,
  girl: NatalMoon,
  options: AshtakootOptions = {},
): AshtakootResult {
  validateNatalMoon(boy, 'boy');
  validateNatalMoon(girl, 'girl');

  const cancellations: string[] = [];

  const koots: KootScore[] = [
    scoreVarna(boy, girl),
    scoreVashya(boy, girl),
    scoreTara(boy, girl),
    scoreYoni(boy, girl),
    scoreGrahaMaitri(boy, girl),
    scoreGana(boy, girl, cancellations, options.ganaCancellation === true),
    scoreBhakoot(boy, girl, cancellations),
    scoreNadi(boy, girl, cancellations),
  ];

  const totalScore = koots.reduce((sum, k) => sum + k.score, 0);
  return { totalScore, koots, cancellations };
}

function validateNatalMoon(m: NatalMoon, label: string): void {
  if (!Number.isInteger(m.rashi) || m.rashi < 0 || m.rashi >= 12) {
    throw new RangeError(`${label}.rashi must be integer in [0, 11], got ${m.rashi}`);
  }
  assertNakshatraIndex(m.nakshatra, `${label}.nakshatra`);
  if (m.lagnaRashi !== undefined &&
      (!Number.isInteger(m.lagnaRashi) || m.lagnaRashi < 0 || m.lagnaRashi >= 12)) {
    throw new RangeError(`${label}.lagnaRashi must be integer in [0, 11], got ${m.lagnaRashi}`);
  }
  if (m.navamsaRashi !== undefined &&
      (!Number.isInteger(m.navamsaRashi) || m.navamsaRashi < 0 || m.navamsaRashi >= 12)) {
    throw new RangeError(`${label}.navamsaRashi must be integer in [0, 11], got ${m.navamsaRashi}`);
  }
  if (m.nakshatraPada !== undefined &&
      (!Number.isInteger(m.nakshatraPada) || m.nakshatraPada < 1 || m.nakshatraPada > 4)) {
    throw new RangeError(`${label}.nakshatraPada must be integer in [1, 4], got ${m.nakshatraPada}`);
  }
}

function scoreVarna(boy: NatalMoon, girl: NatalMoon): KootScore {
  const boyRank = VARNA_RANK[RASHI_VARNA[boy.rashi]!];
  const girlRank = VARNA_RANK[RASHI_VARNA[girl.rashi]!];
  const score = boyRank >= girlRank ? 1 : 0;
  return {
    name: 'Varna',
    score,
    maxScore: 1,
    description: `Boy's varna (${RASHI_VARNA[boy.rashi]}) vs girl's (${RASHI_VARNA[girl.rashi]})`,
  };
}

function scoreVashya(boy: NatalMoon, girl: NatalMoon): KootScore {
  const b = vashyaIndex(RASHI_VASHYA[boy.rashi]!);
  const g = vashyaIndex(RASHI_VASHYA[girl.rashi]!);
  const score = VASHYA_SCORE[b]![g]!;
  return {
    name: 'Vashya',
    score,
    maxScore: 2,
    description: `${RASHI_VASHYA[boy.rashi]} ↔ ${RASHI_VASHYA[girl.rashi]}`,
  };
}

function scoreTara(boy: NatalMoon, girl: NatalMoon): KootScore {
  const dBoyToGirl = ((girl.nakshatra - boy.nakshatra + 27) % 27) + 1;
  const dGirlToBoy = ((boy.nakshatra - girl.nakshatra + 27) % 27) + 1;
  const remBoyToGirl = dBoyToGirl % 9;
  const remGirlToBoy = dGirlToBoy % 9;
  const auspiciousBoyToGirl = !INAUSPICIOUS_TARA_REMAINDERS.includes(remBoyToGirl);
  const auspiciousGirlToBoy = !INAUSPICIOUS_TARA_REMAINDERS.includes(remGirlToBoy);
  const score = (auspiciousBoyToGirl ? 1.5 : 0) + (auspiciousGirlToBoy ? 1.5 : 0);
  return {
    name: 'Tara',
    score,
    maxScore: 3,
    description: `Boy→Girl rem ${remBoyToGirl} (${auspiciousBoyToGirl ? 'auspicious' : 'inauspicious'}); Girl→Boy rem ${remGirlToBoy} (${auspiciousGirlToBoy ? 'auspicious' : 'inauspicious'})`,
  };
}

function scoreYoni(boy: NatalMoon, girl: NatalMoon): KootScore {
  const b = yoniIndex(NAKSHATRA_YONI[boy.nakshatra]!);
  const g = yoniIndex(NAKSHATRA_YONI[girl.nakshatra]!);
  const score = YONI_SCORE[b]![g]!;
  return {
    name: 'Yoni',
    score,
    maxScore: 4,
    description: `${NAKSHATRA_YONI[boy.nakshatra]} ↔ ${NAKSHATRA_YONI[girl.nakshatra]}`,
  };
}

function scoreGrahaMaitri(boy: NatalMoon, girl: NatalMoon): KootScore {
  const boyLord = RASHI_LORD[boy.rashi]!;
  const girlLord = RASHI_LORD[girl.rashi]!;
  let score: number;
  let description: string;
  if (boyLord === girlLord) {
    score = 5;
    description = `Same rashi-lord (graha ${boyLord}), full marks`;
  } else {
    const boyView = NAISARGIKA_MAITRI[boyLord]![girlLord]!;
    const girlView = NAISARGIKA_MAITRI[girlLord]![boyLord]!;
    score = GRAHA_MAITRI_SCORE[maitriIdx(boyView)]![maitriIdx(girlView)]!;
    description = `Boy lord views girl lord as ${maitriLabel(boyView)}; girl lord views boy lord as ${maitriLabel(girlView)}`;
  }
  return { name: 'Graha Maitri', score, maxScore: 5, description };
}

function maitriLabel(v: number): string {
  return v === 1 ? 'friend' : v === -1 ? 'enemy' : 'neutral';
}

function scoreGana(
  boy: NatalMoon,
  girl: NatalMoon,
  cancellations: string[],
  applyCancellation: boolean,
): KootScore {
  const b = ganaIdx(NAKSHATRA_GANA[boy.nakshatra]!);
  const g = ganaIdx(NAKSHATRA_GANA[girl.nakshatra]!);
  let score = GANA_SCORE[b]![g]!;
  let description = `${NAKSHATRA_GANA[boy.nakshatra]} ↔ ${NAKSHATRA_GANA[girl.nakshatra]}`;

  // Doshic cells are only Manushya-Rakshasa (0) and Deva-Rakshasa (1).
  if (applyCancellation && score <= 1) {
    const boyLord = RASHI_LORD[boy.rashi]!;
    const girlLord = RASHI_LORD[girl.rashi]!;
    const sameLord = boyLord === girlLord;
    const mutualFriend =
      NAISARGIKA_MAITRI[boyLord]![girlLord] === 1 &&
      NAISARGIKA_MAITRI[girlLord]![boyLord] === 1;
    if (sameLord || mutualFriend) {
      score = 6;
      const reason = sameLord ? 'same rashi-lord' : 'mutual friendship of rashi-lords';
      description += `, cancelled by ${reason}`;
      cancellations.push(`Gana: ${reason}`);
    }
  }

  return { name: 'Gana', score, maxScore: 6, description };
}

function scoreBhakoot(boy: NatalMoon, girl: NatalMoon, cancellations: string[]): KootScore {
  const dBoyToGirl = ((girl.rashi - boy.rashi + 12) % 12) + 1;
  const dGirlToBoy = ((boy.rashi - girl.rashi + 12) % 12) + 1;

  const isDoshic = BHAKOOT_DOSHIC_DISTANCES.some(
    ([a, b]) => a === dBoyToGirl && b === dGirlToBoy,
  );

  let score = isDoshic ? 0 : 7;
  let description = isDoshic
    ? `Doshic distance (${dBoyToGirl}, ${dGirlToBoy})`
    : `Distance (${dBoyToGirl}, ${dGirlToBoy})`;

  if (isDoshic) {
    const boyLord = RASHI_LORD[boy.rashi]!;
    const girlLord = RASHI_LORD[girl.rashi]!;
    const sameLord = boyLord === girlLord;
    const mutualFriend =
      NAISARGIKA_MAITRI[boyLord]![girlLord] === 1 &&
      NAISARGIKA_MAITRI[girlLord]![boyLord] === 1;
    if (sameLord || mutualFriend) {
      score = 7;
      const reason = sameLord ? 'same rashi-lord' : 'mutual friendship of rashi-lords';
      description += `, cancelled by ${reason}`;
      cancellations.push(`Bhakoot: ${reason}`);
    } else {
      if (boy.lagnaRashi !== undefined && girl.lagnaRashi !== undefined) {
        const boyLagnaLord = RASHI_LORD[boy.lagnaRashi]!;
        const girlLagnaLord = RASHI_LORD[girl.lagnaRashi]!;
        if (boyLagnaLord === girlLagnaLord) {
          score = 7;
          description += ', cancelled by same lagna-lord';
          cancellations.push('Bhakoot: same lagna-lord');
        } else {
          const boySeventhLord = RASHI_LORD[(boy.lagnaRashi + 6) % 12]!;
          const girlSeventhLord = RASHI_LORD[(girl.lagnaRashi + 6) % 12]!;
          if (boySeventhLord === girlSeventhLord) {
            score = 7;
            description += ', cancelled by same 7th-house lord';
            cancellations.push('Bhakoot: same 7th-house lord');
          }
        }
      }
      if (score === 0 && boy.navamsaRashi !== undefined && girl.navamsaRashi !== undefined) {
        const boyNavLord = RASHI_LORD[boy.navamsaRashi]!;
        const girlNavLord = RASHI_LORD[girl.navamsaRashi]!;
        if (boyNavLord === girlNavLord) {
          score = 7;
          description += ', cancelled by same Navamsa lord';
          cancellations.push('Bhakoot: same Navamsa lord');
        }
      }
    }
  }
  return { name: 'Bhakoot', score, maxScore: 7, description };
}

function scoreNadi(boy: NatalMoon, girl: NatalMoon, cancellations: string[]): KootScore {
  const sameNadi = NAKSHATRA_NADI[boy.nakshatra] === NAKSHATRA_NADI[girl.nakshatra];
  let score = sameNadi ? 0 : 8;
  let description = `${NAKSHATRA_NADI[boy.nakshatra]} ↔ ${NAKSHATRA_NADI[girl.nakshatra]}`;

  if (sameNadi) {
    const sameNakshatra = boy.nakshatra === girl.nakshatra;
    const sameRashi = boy.rashi === girl.rashi;
    if (sameNakshatra || sameRashi) {
      score = 8;
      const reason = sameNakshatra ? 'same nakshatra' : 'same rashi';
      description += `, cancelled by ${reason}`;
      cancellations.push(`Nadi: ${reason}`);
    }
  }
  return { name: 'Nadi', score, maxScore: 8, description };
}
